### ISSUE TITLE
Issues with Kibana UAA Auth integration

### ISSUE THREAD

**Issue by mrdavidlaing** [cloudfoundry-community/logsearch-for-cloudfoundry#94]
To implement multi-tenant access, the current Kibana UAA plugin [grabs the orgs and spaces](https://github.com/stayup-io/kibana-cf_authentication/blob/ef2d37d04a8c093ce3bc7481aa9791290e23c46c/index.js#L109-L117) that a user is a member of, and attempts to store them in its (encrypted) session cookie.

This information is then passed extracted and [passed through on every request proxied to Elasticsearch](https://github.com/stayup-io/kibana-cf_authentication/blob/ef2d37d04a8c093ce3bc7481aa9791290e23c46c/index.js#L169-L185); which [Elasticsearch uses to filter the data that is returned](https://github.com/stayup-io/cf-authorization-elasticsearch-plugin/blob/master/src/main/java/io/logsearch/elasticsearch/plugin/CFAuhorizationActionFilter.java#L57-L97).

This works well until the total orgs + spaces membership of the authenticated user exceeds ~ 7.  At this point the size of the session cookie exceeds most browser's cookie length limit; which causes the authentication process to get stuck in a redirect loop.

The workaround is to remove the user from orgs and or spaces until the sum <~ 7.

You can check the number of spaces & orgs the user is a member of using the following cf cli commands:

```
cf login # as admin
cf curl /v2/users | grep -b2 <member-of-many-spaces-username>
cf curl <spaces_url> # see total_results 
cf curl <orgs_url> # see total_results 
```

2 things need to happen to fix this issue.

1.  The Kibana UAA plugin needs to be modified to pass the UAA access token through to Elasticsearch (rather than the list of orgs and spaces)
2.  The Elasticsearch plugin needs to be modified to use the access token to query the UAA directly for a list of authorized orgs and spaces for the user (easy).  This information should be cached at the Elasticsearch level (unknown/hard).

**Comment by mrdavidlaing** [comment]
PIvotal have decided not to support multi-tenant access to Elasticsearch; so Pivotal engineers won't be contributing the above suggested code fix.

I'm be happy to support anyone who wants to pick up and implement this feature - just drop me an email at dlaing@pivotal.io

**Comment by MaheshRudrachar** [comment]
Hi mrdavidlaing,

I am in process of utilizing kibana-cf-authentication plugin for multt-tenant filtering access to elasticsearch. I found few issues in elastic-search plugin but got it fixed related to TermQuery builder. As you correctly mentioned, now I am stuck in redirection loop because user i.e, Admin/User having more Orgs and spaces than expected.

I would need your help in implementing the workaround you have suggested.
1) UAA token can be passed to Elasticsearch using prehandle event that has been implemented right.
Basically no need to callback for /orgs and /spaces and get the oauth token then pass it to elasticsearch as header along with user profile.

2) In elasticsearch instead of registering to X-Orgs and X-spaces, we need to call back UAA using the access token that was passed.. Is that right

Do you have sample or stubs for these, that would be helpful....

**Comment by mrdavidlaing** [comment]
@MaheshRudrachar - I'm sorry; I don't have any example code I can share.

However, the process you have outlined sounds exactly right for me.

We're going to be moving this repo over to cloudfoundry-community this week; at which time I'll submit a PR containing the locations where I think the code should go which should give you a place to start.

**Comment by MaheshRudrachar** [comment]
Sure. Thanks a lot...

**Comment by sharms** [comment]
Why not merge https://github.com/stayup-io/kibana-cf_authentication/pull/1 ? This is already solved for a few months --

**Comment by Infra-Red** [comment]
Fixed in the https://github.com/cloudfoundry-community/logsearch-for-cloudfoundry/pull/112 PR, @axelaris i think we can close this issue.
